File: src/py/enumeration.py
from collections import Counter
import argparse
from typing import Union

Vec = tuple[int, int, int, int, int, int]

tup_bank = {}
primes = (2, 3, 5, 7, 11, 13, 17, 19, 23)


def gen_tup(s: int, n: int) -> list[tuple[int, ...]]:
    if (s, n) in tup_bank.keys():
        return tup_bank[(s, n)]

    if s == 0:
        ret = [(0,) * n]
        tup_bank[(s, n)] = ret
        return ret

    elif n == 1:
        ret = [(s,)]
        tup_bank[(s, n)] = ret
        return ret

    out = []
    for i in range(0, s + 1):
        prev = gen_tup(s - i, n - 1)
        out += [(i,) + p for p in prev]
    tup_bank[(s, n)] = out
    return out


def gen_rows(pows: tuple[int, ...], n: int) -> set[Vec]:
    rows = {(1,) * n}
    for p, pow in zip(primes, pows):
        ptups = gen_tup(pow, n)
        new_rows = set()
        for row in rows:
            for tup in ptups:
                new_rows.add(
                    tuple(
                        sorted(
                            [elt * (p**r) for elt, r in zip(row, tup)], reverse=True
                        )
                    )
                )
        rows = new_rows
    return rows


def split_by_sum(all_rows: set[Vec]) -> list[tuple[int, int]]:
    counts = {}
    unique_counts = {}
    for row in all_rows:
        s = sum(row)
        counts[s] = counts.get(s, 0) + 1
        if len(row) == len(set(row)):
            unique_counts[s] = unique_counts.get(s, 0) + 1

    ucs = sorted(unique_counts.items(), key=lambda x: x[1], reverse=True)
    return ucs


def rows_to_rowdict(all_rows: set[Vec]) -> dict[int, list[Vec]]:
    ucs = split_by_sum(all_rows)
    S_set = {S for S, c in ucs if c >= 12}
    row_dict = {S: [] for S in S_set}

    print(f">>> (enum) total possible 6-vecs: {sum(u[1] for u in ucs)}")
    for row in all_rows:
        if len(set(row)) == len(row) and sum(row) in S_set:
            row_dict[sum(row)].append(row)

    # reduction: if there's less than 12 vecs, we can't fill the grid
    for S, vecs in row_dict.items():
        while len(vecs) >= 12:
            count = Counter([elt for vec in vecs for elt in vec])
            if count.most_common()[-1][1] >= 2:
                break
            # reduction: we can't use vecs that have unique elts
            vecs = [vec for vec in vecs if all(count[elt] > 1 for elt in vec)]
        else:
            vecs = []
        row_dict[S] = vecs

    row_dict = {S: vecs for S, vecs in row_dict.items() if vecs}

    return row_dict


def write_enums(row_dict: dict[int, list[Vec]], file: str, SS: Union[list[int], None] = None):
    # rows are guaranteed to written be in ascending order of sum
    n_rows = 0
    with open(file, "w") as f:
        for s in sorted(row_dict.keys()):
            if (not SS) or (s in SS):
                for row in row_dict[s]:
                    print(" ".join(str(i) for i in row), file=f)
                    n_rows += 1
    print(f">>> wrote {n_rows} rows to {file} in ascending order of sum")

def write_counts(row_dict: dict[int, list[Vec]], file: str, SS: Union[list[int], None] = None):
    counts = {}
    for s in sorted(row_dict.keys()):
        if (not SS) or (s in SS):
            counts[s] = len(row_dict[s])
    with open(file, "w") as f:
        for s, c in sorted(counts.items(), key=lambda x: x[1], reverse=True):
            print(f"{s} {c}", file=f)
    print(f">>> wrote counts to {file}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--file", metavar="file", type=str)
    parser.add_argument("--sum", metavar="S", type=int, nargs="*")
    parser.add_argument("-c", action="store_true")
    parser.add_argument("exponents", metavar="P", type=int, nargs="+")
    args = parser.parse_args()

    P = tuple(args.exponents)
    N = 6
    SS = args.sum
    all_rows = gen_rows(P, N)
    row_dict = rows_to_rowdict(all_rows)
    if args.c:
        write_counts(row_dict, args.file, SS)
    else:
        write_enums(row_dict, args.file, SS)

File: src/c/enumeration.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

typedef std::vector<size_t> vec_t;
typedef std::vector<vec_t> vecs_t;

std::map<std::pair<size_t, size_t>, vecs_t> tup_bank;

vecs_t &gen_tup(size_t s, size_t n) {
  auto it = tup_bank.find(std::make_pair(s, n));
  if (it != tup_bank.end()) {
    return it->second;
  }
  tup_bank[std::make_pair(s, n)] = vecs_t();
  auto &res = tup_bank[std::make_pair(s, n)];
  if (s == 0) {
    vec_t new_p(n);
    std::fill(new_p.begin(), new_p.end(), 0);
    res.push_back(new_p);
  } else if (n == 1) {
    res.push_back({s});
  } else {
    for (size_t i = 0; i <= s; i++) {
      for (auto p : gen_tup(s - i, n - 1)) {
        auto new_p(p);
        new_p.push_back(i);
        res.push_back(new_p);
      }
    }
  }
  return res;
}

typedef std::multiset<size_t> row_t;
typedef std::set<row_t> rows_t;

const size_t VEC_SIZE = 6;
const vec_t PRIMES{2, 3, 5, 7, 11, 13, 17, 19, 23};

row_t read_row(std::ifstream &rows) {
  row_t row;
  for (size_t i = 0; i < VEC_SIZE; i++) {
    size_t elt;
    rows >> elt;
    row.insert(elt);
  }
  return row;
}

void write_row(std::ofstream &rows, row_t &row) {
  for (auto elt : row) {
    rows << elt << " ";
  }
}

// returns filename where rows are written
char *gen_rows(vec_t &pows, char *rows_filename, char *new_rows_filename) {
  // setup
  {
    std::ofstream rows;
    rows.open(rows_filename);
    for (size_t i = 0; i < VEC_SIZE; i++) {
      rows << 1 << " ";
    }
    rows.close();
  }

  // main loop
  for (size_t i = 0; i < pows.size(); i++) {
    auto p = PRIMES[i];
    auto pow = pows[i];
    auto ptups = gen_tup(pows[i], VEC_SIZE);

    std::ifstream rows;
    rows.open(rows_filename);
    std::ofstream new_rows;
    new_rows.open(new_rows_filename);

    while (!rows.eof()) {
      auto row = read_row(rows);
      for (auto tup : ptups) {
        row_t new_row;
        auto r = tup.begin();
        for (auto elt : row) {
          auto new_elt = elt;
          for (size_t j = 0; j < *r; j++) {
            new_elt *= p;
          }
          new_row.insert(new_elt);
          r++;
        }
        write_row(new_rows, new_row);
      }
    }

    rows.close();
    new_rows.close();
    std::swap(rows_filename, new_rows_filename);
  }

  tup_bank.clear();

  // remove rows with duplicates
  {
    std::ifstream rows;
    rows.open(rows_filename);
    std::ofstream new_rows;
    new_rows.open(new_rows_filename);

    while (!rows.eof()) {
      auto row = read_row(rows);
      if (std::adjacent_find(row.begin(), row.end()) != row.end()) {
        continue;
      }
      write_row(new_rows, row);
    }

    rows.close();
    new_rows.close();
    std::swap(rows_filename, new_rows_filename);
  }

  return rows_filename;
}

std::map<size_t, size_t> split_by_sum(char *rows_filename, char *out_filename) {
  std::map<size_t, size_t> row_counts;

  std::ifstream all_rows;
  std::ofstream out_rows;
  all_rows.open(rows_filename);
  out_rows.open(out_filename);
  while (!all_rows.eof()) {
    auto row = read_row(all_rows);
    if (std::adjacent_find(row.begin(), row.end()) != row.end()) {
      continue;
    }
    size_t sum = 0;
    for (auto elt : row) {
      sum += elt;
    }
    if (row_counts.find(sum) == row_counts.end()) {
      row_counts[sum] = 0;
    }
    row_counts[sum]++;
    write_row(out_rows, row);
  }
  all_rows.close();
  out_rows.close();

  size_t total = 0;
  for (auto pair : row_counts) {
    total += pair.second;
  }
  std::cout << ">>> (enum) total possible 6-vecs: " << total << "\n";

  return row_counts;
}

typedef std::map<size_t, std::vector<row_t>> rowdict_t;

// rowdict_t rows_to_rowdict(char *rows_filename, char *out_filename) {
//   rowdict_t rowdict;

//   auto row_counts = split_by_sum(rows_filename, out_filename);

//   auto rows_it = row_counts.begin();
//   while (rows_it != row_counts.end()) {
//     auto S = rows_it->first;
//     auto row_count = rows_it->second;
//     if (row_count < 12) {
//       rows_it = row_counts.erase(rows_it);
//       continue;
//     }
//   }

//   // reduction
//   auto rows_it = rowdict.begin();
//   while (rows_it != rowdict.end()) {
//     auto S = rows_it->first;
//     auto rows = rows_it->second;

//     // if < 12, remove and continue
//     if (rows.size() < 12) {
//       rows_it = rowdict.erase(rows_it);
//       continue;
//     }

//     // counter
//     std::map<size_t, size_t> count;
//     for (auto row : rows) {
//       for (auto elt : row) {
//         if (count.find(elt) == count.end()) {
//           count[elt] = 0;
//         }
//         count[elt]++;
//       }
//     }

//     // remove while some count[elt] is 1
//     bool did_remove = true;
//     while (did_remove && rows.size() >= 12) {
//       did_remove = false;
//       auto it = rows.begin();
//       while (it != rows.end()) {
//         if (std::any_of(it->begin(), it->end(),
//                         [&count](size_t elt) { return count[elt] == 1; })) {
//           for (auto elt : *it) {
//             count[elt]--;
//           }
//           did_remove = true;
//           it = rows.erase(it);
//         } else {
//           it++;
//         }
//       }
//     }

//     // if < 12, remove and continue
//     if (rows.size() < 12) {
//       rows_it = rowdict.erase(rows_it);
//       continue;
//     } else {
//       rows_it->second = rows;
//       rows_it++;
//     }
//   }

//   return rowdict;
// }

int main(int argc, char *argv[]) {
  char *filename = argv[1];
  vec_t exponents(argc - 2);

  for (size_t i = 0; i < exponents.size(); i++) {
    exponents[i] = std::stol(argv[i + 2], NULL, 10);
  }

  char x[5] = "test";
  char y[6] = "test2";
  gen_rows(exponents, x, y);

  // auto rows = rows_to_rowdict(gen_rows(exponents));

  // // write rows
  // std::ofstream file;
  // file.open(filename);

  // for (auto vv : rows) {
  //   for (auto v : vv.second) {
  //     auto it = v.rbegin();
  //     file << *it++;
  //     while (it != v.rend()) {
  //       file << " " << *it++;
  //     }
  //     file << "\n";
  //   }
  // }

  // std::cout << ">>> (enum) wrote " << rows.size() << " rows to " << filename
  //           << " in ascending order of sum\n";
}
